File: src/agentflow_computer_mcp/cli/rest_client.py
"""REST client for `agentflow.website/_agents` with `x-api-key` auth.

Read order for the api key:
  1. explicit `api_key=` argument
  2. AGENTFLOW_API_KEY env
  3. ~/.agentflow/auth.json
"""
from __future__ import annotations

import os
from typing import Any

import httpx

from ..config import load_auth

DEFAULT_BASE = "https://agentflow.website/_agents"


class NotAuthenticated(RuntimeError):
    """No api key found in flag, env, or auth.json."""


class ServerError(RuntimeError):
    """Non-2xx HTTP response."""

    def __init__(self, status: int, body: str) -> None:
        super().__init__(f"server returned {status}: {body[:200]}")
        self.status = status
        self.body = body


def resolve_api_key(explicit: str | None = None) -> str:
    if explicit:
        return explicit
    env = os.environ.get("AGENTFLOW_API_KEY") or os.environ.get("AF_API_KEY")
    if env:
        return env
    auth = load_auth()
    if auth.api_key:
        return auth.api_key
    raise NotAuthenticated("не авторизован; запусти `agentflow login`")


def _client(api_key: str, base: str) -> httpx.Client:
    return httpx.Client(
        base_url=base,
        headers={"x-api-key": api_key, "user-agent": "agentflow-cli/1"},
        timeout=15.0,
    )


def _check(resp: httpx.Response) -> Any:
    if resp.status_code >= 400:
        raise ServerError(resp.status_code, resp.text)
    if not resp.content:
        return None
    return resp.json()


def get(path: str, *, api_key: str | None = None, base: str = DEFAULT_BASE,
        params: dict[str, Any] | None = None) -> Any:
    key = resolve_api_key(api_key)
    with _client(key, base) as c:
        return _check(c.get(path, params=params))


def post(path: str, json_body: dict[str, Any], *, api_key: str | None = None,
         base: str = DEFAULT_BASE) -> Any:
    key = resolve_api_key(api_key)
    with _client(key, base) as c:
        return _check(c.post(path, json=json_body))

File: macapp/Sources/AgentFlow/RestClient.swift
import Foundation

public enum RestError: Error, Equatable {
    case notAuthenticated
    case http(Int, String)
    case decode(String)
    case network(String)
}

/// REST client for `https://agentflow.website/_agents`.
public final class RestClient {
    public static let defaultBase = URL(string: "https://agentflow.website/_agents")!

    private let base: URL
    private let session: URLSession
    private let apiKeyProvider: () -> String?

    public init(
        base: URL = RestClient.defaultBase,
        session: URLSession = .shared,
        apiKey: @escaping () -> String?
    ) {
        self.base = base
        self.session = session
        self.apiKeyProvider = apiKey
    }

    /// Convenience: build a default-config client reading `~/.agentflow/auth.json`.
    public static func defaultClient() -> RestClient {
        RestClient { AuthLoader.load()?.apiKey }
    }

    // MARK: - public API

    public func fetchGoals() async throws -> [Goal] {
        let data = try await get(path: "/me/autonomous/goals")
        // The server returns either `{"goals":[...]}` or `[...]`. Handle both.
        if let env = try? JSONDecoder().decode(GoalsEnvelope.self, from: data),
           let goals = env.goals ?? env.items ?? env.results {
            return goals
        }
        if let arr = try? JSONDecoder().decode([Goal].self, from: data) {
            return arr
        }
        throw RestError.decode("goals payload did not match list or envelope shape")
    }

    public func fetchBudget() async throws -> Budget {
        let data = try await get(path: "/me/autonomous/budget")
        do {
            return try JSONDecoder().decode(Budget.self, from: data)
        } catch {
            throw RestError.decode("budget: \(error)")
        }
    }

    // MARK: - transport

    func get(path: String) async throws -> Data {
        guard let key = apiKeyProvider(), !key.isEmpty else {
            throw RestError.notAuthenticated
        }
        let url = base.appendingPathComponent(path.hasPrefix("/") ? String(path.dropFirst()) : path)
        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        req.setValue(key, forHTTPHeaderField: "x-api-key")
        req.setValue("agentflow-macapp/1", forHTTPHeaderField: "user-agent")
        req.timeoutInterval = 15

        let (data, resp): (Data, URLResponse)
        do {
            (data, resp) = try await session.data(for: req)
        } catch {
            throw RestError.network(String(describing: error))
        }
        guard let http = resp as? HTTPURLResponse else {
            throw RestError.network("non-HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw RestError.http(http.statusCode, body)
        }
        return data
    }
}
